prboom2/src/gl_shader.c

```diff
@@ -53,217 +54,534 @@
-} shdr_indexed_unif_t;
+  const GLchar* name;
+  const GLchar* value;
+} shdr_def_t;
 
-// Fuzz shader uniform bindings
-typedef struct shdr_fuzz_unif_s
+typedef struct shdr_src_s
 {
-  // (vec2) sprite texture dimensions
-  int tex_d_index;
-  // (float) ratio of screen resolution to fuzz resolution
-  int ratio_index;
-  // (float) random seed
-  int seed_index;
-} shdr_fuzz_unif_t;
+  GLchar const** strs;
+  GLint* lens;
+  unsigned int size;
+  unsigned int cap;
+} shdr_src_t;
 
-static GLShader *sh_indexed = NULL;
-static shdr_indexed_unif_t indexed_unifs;
-static GLShader *sh_fuzz = NULL;
-static shdr_fuzz_unif_t fuzz_unifs;
-static GLShader *active_shader = NULL;
-
-static GLShader* gld_LoadShader(const char *vpname, const char *fpname);
+static void ShaderSrcInit(shdr_src_t* src)
+{
+  memset(src, 0, sizeof(*src));
+}
 
-static void get_indexed_shader_bindings()
+static void ShaderSrcDestroy(shdr_src_t* src)
 {
-  int idx;
+  if (src->strs)
+    Z_Free(src->strs);
+  if (src->lens)
+    Z_Free(src->lens);
+  ShaderSrcInit(src);
+}
 
-  indexed_unifs.lightlevel_index = GLEXT_glGetUniformLocationARB(sh_indexed->hShader, "lightlevel");
+static void ShaderSrcAppend(shdr_src_t* src, const GLchar* str, GLint len)
+{
+  unsigned int size = src->size;
 
-  GLEXT_glUseProgramObjectARB(sh_indexed->hShader);
+  if (size && src->lens[size - 1] >= 0 && src->strs[size - 1] + src->lens[size - 1] == str)
+  {
+    // Fast path -- expand last string to include more of source buffer
+    src->lens[src->size - 1] += len;
+    return;
+  }
 
-  idx = GLEXT_glGetUniformLocationARB(sh_indexed->hShader, "tex");
-  GLEXT_glUniform1iARB(idx, 0);
+  if (src->size == src->cap)
+  {
+    if (src->cap == 0)
+      src->cap = 1024;
+    else
+      src->cap *= 2;
 
-  idx = GLEXT_glGetUniformLocationARB(sh_indexed->hShader, "colormap");
-  GLEXT_glUniform1iARB(idx, 2);
+    src->strs = Z_Realloc(src->strs, src->cap * sizeof(*src->strs));
+    src->lens = Z_Realloc(src->lens, src->cap * sizeof(*src->lens));
+  }
 
-  GLEXT_glUseProgramObjectARB(0);
+  src->strs[src->size] = str;
+  src->lens[src->size] = len;
+  ++src->size;
 }
 
-static void get_fuzz_shader_bindings()
+static void ShaderLookup(const char* name, GLchar const** text, GLint* len)
 {
-  int idx;
+  int lump = W_CheckNumForName2(name, ns_prboom);
 
-  fuzz_unifs.tex_d_index = GLEXT_glGetUniformLocationARB(sh_fuzz->hShader, "tex_d");
-  fuzz_unifs.ratio_index = GLEXT_glGetUniformLocationARB(sh_fuzz->hShader, "ratio");
-  fuzz_unifs.seed_index = GLEXT_glGetUniformLocationARB(sh_fuzz->hShader, "seed");
+  if (lump == LUMP_NOT_FOUND)
+    I_Error("Could not find shader source: %s\n", name);
 
-  GLEXT_glUseProgramObjectARB(sh_fuzz->hShader);
+  *text = W_LumpByNum(lump);
+  *len = W_LumpLength(lump);
+}
 
-  idx = GLEXT_glGetUniformLocationARB(sh_fuzz->hShader, "tex");
-  GLEXT_glUniform1iARB(idx, 0);
+static dboolean GetLine(GLchar const** text, GLint* len, GLchar const** line, GLint* llen)

```

Comment:
I don't think this file should have custom string handling code. We have `dsda_SplitString(text, "\n\r")` to get the text line by line (e.g., in `dsda_LoadWadPreferences`). If something more robust is required, a separate parser file using the scanner might be safer (see `udmf.cpp` for an example).

Reply:
`dsda_SplitString` is kind of heavy:

- The entire shader lump has to be copied and NUL-terminated before calling it
- It allocates and splits in one go instead of iterating over the string incrementally
- It modifies the string in place (via `strtok`), so the `\n` needs to be added back to each line before passing it to `glShaderSource`.
- `glShaderSource` accepts strings as pointer + length (no NUL required) and does a final copy and concatenation internally, so this is all wasted

It's fine for conveniently splitting small strings, but a file parser needs a different architecture.

I don't really want to use a lexical analyzer on the source because I'm trying to interpret as little GLSL syntax as possible.  I just want to expand `#include`s and insert `#define`s in the right place.

Reply:
Even if the shader was thousands of lines, would it be measurable? :^)

Reply:
It would be a bunch of extra bookkeeping in the rest of the code (allocs, copies, cleanup) just to get rid of the `GetLine` function, which is extremely simple.  I think it would be better for the utility library to grow pointer/length string functions instead.  NUL-terminated strings are just not very flexible, e.g. you can't create a substring without mutating or allocating and copying.

Reply:
That would also be fine with me - basically if there are missing utility tasks that aren't specific to the shader then it would be best to create them separately. The caveat would be things that the scanners already do, because I want to avoid having multiple different implementations of the same task.

Reply:
The scanner isn't a good match because it's both overkill (I don't need to parse the vast majority of token types, because I'm doing very light preprocessing of the source), but doesn't handle things I do need (C preprocessor-style directives).  I'll see about adding the start of a pointer/length API to the DSDA utilities.